src/core/tsi/ssl_transport_security.cc

```diff
@@ -853,6 +882,38 @@ static int NullVerifyCallback(int /*preverify_ok*/, X509_STORE_CTX* /*ctx*/) {
   return 1;
 }
 
+static int RootCertExtractCallback(int preverify_ok, X509_STORE_CTX* ctx) {
+  // There's a case where this function is set in SSL_CTX_set_verify and a CRL
+  // related callback is set with X509_STORE_set_verify_cb. They overlap and
+  // this will take precedence, thus we need to ensure the CRL related callback
+  // is still called
+  X509_VERIFY_PARAM* param = X509_STORE_CTX_get0_param(ctx);
+  auto flags = X509_VERIFY_PARAM_get_flags(param);
+  if (flags & X509_V_FLAG_CRL_CHECK) {
+    preverify_ok = verify_cb(preverify_ok, ctx);
+  }
+
+  if (preverify_ok == 0) {
+    return 0;
+  }
+
+  // If we're here, verification was successful
+  // Get the verified chain from the X509_STORE_CTX and put it on the SSL object
+  // so that we have access to it when populating the tsi_peer
+  STACK_OF(X509)* chain = X509_STORE_CTX_get0_chain(ctx);
+  if (chain == nullptr) {
+    return 1;
```

Comment:
Shouldn't all return values continue to be `preverify_ok`? We don't want to change the existing behavior, just opportunistically extract this info.

Reply:
Above this is a line for `if preverify_ok == 0 return 0;`, so behaviorally here `preverify_ok` will always be 1, but it could be cleaner to just re-use `preverify_ok`.
One of the intents here was that, if `preverify_ok` is _not_ ok (`== 0`) then there won't be a verified chain so we should not continue.

Thoughts?

Reply:
Sounds reasonable.